Show errors in rate step

v18 opened the pull request:
# Changes in this PR
- check rates prop exists before using it in the shipping-label reducer
- add server errors to `getRatesErrors` selector
- if there's an error sent from the server for a package's rates, show a `FieldError`
- add Enzyme tests for `<ShippingRates />`

# Testing
Check out the [counterpart PR on the server](https://github.com/Automattic/woocommerce-connect-server/pull/802). Use the testing instructions there to simulate different responses to see the errors show up on the rates step. You can trigger a re-fresh of the rates by re-saving your packages on the modal :)

# What does it look like?

When there are no rates, only errors:

<img src="https://cloud.githubusercontent.com/assets/11487924/25467225/e964f988-2ada-11e7-971a-c911e3b56b4e.png" />

Where there is both a rate and an error (we don't currently support sending down both. if there are errors and rates we'll send only the rates - but there is good stylin' in just in case that situation ever happens):

<img src="https://cloud.githubusercontent.com/assets/11487924/25467257/103b86da-2adb-11e7-8d2b-cb557cd0aaef.png">

# Stacked PRs!
The following PRs are related and stacked:
https://github.com/Automattic/woocommerce-services/pull/997
https://github.com/Automattic/woocommerce-services/pull/998
https://github.com/Automattic/woocommerce-services/pull/999
https://github.com/Automattic/woocommerce-services/pull/1000

jeffstieler commented on client/shipping-label/views/purchase/steps/rates/test/list.js:
```diff
@@ -0,0 +1,384 @@
+			} );
+			it( 'should display error', () => {
+				const errorWrapper = wrapper.find( FieldError );
+				// const errorText = errorWrapper.text();
```
Please remove this line.

v18 replied:
👍 Removed in 08a888b935617c03f6a7f0c21e826b36921a84c3

jeffstieler commented on client/shipping-label/views/purchase/steps/rates/test/list.js:
```diff
@@ -0,0 +1,384 @@
+			} );
+			it( 'should display extra heading for package title', () => {
+				const packageNames = wrapper.find( 'p' ).map( node => node.text() );
+				expect( wrapper.find( 'p' ) ).to.have.length( 2 );
```
Can another `.find()` call be avoided here by reusing the `packageNames` variable? Or perhaps can this line be omitted in favor of the line below?

v18 replied:
Yes, fixed in e35b2109ab925a4de55c2ae7fdd94acec3d7efac

jeffstieler commented on client/shipping-label/views/purchase/steps/rates/test/list.js:
```diff
@@ -0,0 +1,384 @@
+			} );
+		} );
+
+		describe( 'Both rates and errors for all packages', () => {
```
Can we add a test case for only one package (of multiple) having errors?

v18 replied:
Sure, I can add it, but it seems a bit superfluous to me. We're already testing the display of these cases:
- multiple packages
- errors only for each package
- rates only for each package
- both rates and errors for a package

Since we are capable of displaying multiple packages, and for each package we are capable of displaying each type: both errors and rates, rates only, and errors only, then in terms of display, we're all set.

What do you think?

In any case, I did some of the groundwork and it's available in 8c0b08ef000df3dd28e3d97842402e87fd04307c

jeffstieler commented on client/shipping-label/state/selectors/errors.js:
```diff
@@ -58,7 +58,22 @@ const getPackagesErrors = ( values ) => _.mapValues( values, ( pckg ) => {
+					return {
+						message: ( error.userMessage ||
+							error.message ||
+							"We couldn't get a rate for this package, please try again." ),
```
Missing translation.

v18 replied:
Fixed in 15a2063739c7f756a31f77aa3d6a13aefd3fa8d0

v18 commented on client/shipping-label/state/selectors/errors.js:
```diff
@@ -58,7 +58,22 @@ const getPackagesErrors = ( values ) => _.mapValues( values, ( pckg ) => {
+const getRatesErrors = ( { values: selectedRates, available: allRates } ) => {
+	return {
+		server: _.mapValues( allRates, ( rate ) => {
+			if ( rate.errors ) {
```
Yes, it's intended - the form errors return value will be the same as before, and the server return value will be either an array containing all the errors, or an empty array.

I added tests so that the behaviour is clarified in 3e0276feab8f6625fbf33e42a0dc8d7b30259ca8. Also tests 🎉

Note that I also changed the errors so that we're only getting the error message, and not the entire object (related to your next comment).

v18 commented on client/shipping-label/state/selectors/errors.js:
```diff
@@ -58,7 +58,22 @@ const getPackagesErrors = ( values ) => _.mapValues( values, ( pckg ) => {
+						message: ( error.userMessage ||
+							error.message ||
+							"We couldn't get a rate for this package, please try again." ),
+					};
```
You're right, should just be a plain error string. Fixed in 189f98ef79af32d59791a90ae8ae9b0afa6447da

v18 commented on client/shipping-label/state/selectors/errors.js:
```diff
@@ -58,7 +58,22 @@ const getPackagesErrors = ( values ) => _.mapValues( values, ( pckg ) => {
+					};
+				} );
+			}
+		} ),
```
Challenge accepted 😉 One indentation level was removed when using a string instead of an error object, and I removed two more in 15a2063739c7f756a31f77aa3d6a13aefd3fa8d0

v18 commented on client/shipping-label/views/purchase/steps/rates/test/list.js:
```diff
@@ -0,0 +1,384 @@
+			} );
+			it( 'should not display extra heading for package title', () => {
+				// the package title is displayed using the dropdown legend
+				expect( wrapper.find( 'p' ) ).to.have.length( 0 );
```
Sure - fixed in e35b2109ab925a4de55c2ae7fdd94acec3d7efac

v18 commented:
@jeffstieler thanks for the review. I fixed all the issues you mentioned, just had a comment on the extra test comment: https://github.com/Automattic/woocommerce-services/pull/1001#discussion_r117020597

v18 commented:
Removed the unneeded test code, and squashed. Retested on the old and new server versions (i.e. with and without sending errors in the response). Tests pass. Merging in now.
